editor/export/export_template_manager.cpp

```diff
@@ -655,39 +656,78 @@ void ExportTemplateManager::_hide_dialog() {
+	if (p_preset.is_valid()) {
+		String android_source_zip = p_preset->get("gradle_build/android_source_template");
+		if (!android_source_zip.is_empty()) {
+			return android_source_zip;
+		}
+	}
+
 	const String templates_dir = EditorPaths::get_singleton()->get_export_templates_dir().path_join(VERSION_FULL_CONFIG);
-	return FileAccess::exists(templates_dir.path_join("android_source.zip"));
+	return templates_dir.path_join("android_source.zip");
 }
 
-Error ExportTemplateManager::install_android_template() {
-	const String &templates_path = EditorPaths::get_singleton()->get_export_templates_dir().path_join(VERSION_FULL_CONFIG);
-	const String &source_zip = templates_path.path_join("android_source.zip");
+String ExportTemplateManager::get_android_template_identifier(const Ref<EditorExportPreset> &p_preset) {
+	// The template identifier is the version for the default template, and the full path for custom templates.
+	if (p_preset.is_valid()) {
+		String android_source_zip = p_preset->get("gradle_build/android_source_template");
+		if (!android_source_zip.is_empty()) {
+			return android_source_zip + String(" [") + FileAccess::get_md5(android_source_zip) + String("]");
+		}
+	}
+	return VERSION_FULL_CONFIG;
+}
+
+bool ExportTemplateManager::is_android_template_installed(const Ref<EditorExportPreset> &p_preset) {
+	return DirAccess::exists(get_android_build_directory(p_preset));
+}
+
+bool ExportTemplateManager::can_install_android_template(const Ref<EditorExportPreset> &p_preset) {
+	return FileAccess::exists(get_android_source_zip(p_preset));
+}
+
+Error ExportTemplateManager::install_android_template(const Ref<EditorExportPreset> &p_preset) {
+	const String source_zip = get_android_source_zip(p_preset);
 	ERR_FAIL_COND_V(!FileAccess::exists(source_zip), ERR_CANT_OPEN);
-	return install_android_template_from_file(source_zip);
+	return install_android_template_from_file(source_zip, p_preset);
 }
-Error ExportTemplateManager::install_android_template_from_file(const String &p_file) {
+
+Error ExportTemplateManager::install_android_template_from_file(const String &p_file, const Ref<EditorExportPreset> &p_preset) {
 	// To support custom Android builds, we install the Java source code and buildsystem
 	// from android_source.zip to the project's res://android folder.
 
 	Ref<DirAccess> da = DirAccess::create(DirAccess::ACCESS_RESOURCES);
 	ERR_FAIL_COND_V(da.is_null(), ERR_CANT_CREATE);
 
-	// Make res://android dir (if it does not exist).
-	da->make_dir("android");
+	String build_dir = get_android_build_directory(p_preset);
+	String parent_dir = build_dir.get_base_dir();
+
+	// Make parent of the build dir (if it does not exist).
+	da->make_dir_recursive(parent_dir);
 	{
-		// Add version, to ensure building won't work if template and Godot version don't match.
-		Ref<FileAccess> f = FileAccess::open("res://android/.build_version", FileAccess::WRITE);
+		// Add identifier, to ensure building won't work current template doesn't match.
```

Comment:
Seems like they are missing words between `work` and `current template`.

Reply:
Heh, yes, you're right :-) Fixed!